### ISSUE TITLE
ErrorDomain Code=-999 - Requests gets auto-canceled when request made on notification click from background

### ISSUE THREAD

**Issue by Hamna21** [Alamofire/Alamofire#3625]
## What did you do?

I am making a GET call to the server when the user clicks on a notification in the background.
I have a singleton Shared Session which I use for all my network requests. 

`struct NetworkClientRetrier: RequestInterceptor {
		func retry(_ request: Request, for session: Session, dueTo error: Error, completion: @escaping (RetryResult) -> Void) {
			if let response = request.task?.response as? HTTPURLResponse,
			   response.statusCode >= 500,
			   request.retryCount < 2 {
				completion(.retryWithDelay(5))
			} else {
				completion(.doNotRetryWithError(error))
			}
		}
		
		func adapt(_ urlRequest: URLRequest, for session: Session, completion: @escaping (Swift.Result<URLRequest, Error>) -> Void) {
			Auth.auth().currentUser?.getIDTokenForcingRefresh(false) { idToken, error in
				guard error == nil, let token = idToken else {
					return
				}
				var adaptedRequest = urlRequest
				adaptedRequest.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
				completion(.success(adaptedRequest))
			}
		}
	}`

`static let APIManager: Session = {
		let configuration = URLSessionConfiguration.default
		configuration.timeoutIntervalForRequest = 120
		configuration.httpMaximumConnectionsPerHost = 20
		let delegate = Session.default.delegate
		let manager = Session.init(configuration: configuration,
								   delegate: delegate,
								   startRequestsImmediately: true,
								   interceptor: NetworkClientRetrier())
		return manager
	}()`

On the click of notification, I make multiple network requests when app becomes active.

ℹ Please replace this with what you expected to happen.  
I expected that the call is not auto-canceled.

ℹ Please replace this with of what happened instead.  

The requests are not always successful and get auto-canceled. I get the following error even after using a singleton:

`originalError: Alamofire.AFError.sessionTaskFailed(error: Error Domain=NSURLErrorDomain Code=-999 "cancelled" UserInfo={_kCFStreamErrorCodeKey=89, NSUnderlyingError=0x12e671790 {Error Domain=kCFErrorDomainCFNetwork Code=-999 "(null)" UserInfo={NSErrorPeerAddressKey=<CFData 0x115c740e0 [0x21bee81b8]>{length = 16, capacity = 16, bytes = 0x100201bbd8ef26350000000000000000}, _kCFStreamErrorCodeKey=89, _kCFStreamErrorDomainKey=1}}, _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <EEC98F74-2DAE-4937-9559-20204ADFB8DF>.<16>, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <EEC98F74-2DAE-4937-9559-20204ADFB8DF>.<16>"
),`

## Alamofire Environment

**Alamofire Version:** : 5.4.4
**Xcode Version:**: 13.4.1

**Comment by jshier** [comment]
Please describe the sequence of events more clearly. What notification is being tapped, one for your app? You're making requests when the app becomes active? What happens to the app after that? Is it becoming active in the background or the foreground?

If this is happening when your app is activated in the background, you need to looking into background tasks: https://developer.apple.com/forums/thread/85066

**Comment by jshier** [comment]
Also, there's a few other things wrong with your code.

- You can use `validate()` to produce a status code error rather than having to dig into `task.request`.
- You can probably use one of the built in retriers rather than your own for stuff like that.
- Your adapter doesn't properly call the completion handler when there's an error, meaning the request will never complete.
- You can use `adaptedRequest.headers` to set the authorization header.

**Comment by Hamna21** [comment]
Chain: View Controller -> Network Request -> Network Client 
`View controller` calls this function in the `Network Request` class:
```swift
func getPersonalUserInsight(insightId: Int,
								completionHandler: @escaping (Swift.Result<PersonalUserInsight, Error>) -> Void){
		firstly {
			NetworkClient.requestObject(NetworkRouter.insight(insightId: insightId), type: PersonalUserInsight.self)
		}.done { (response) in
			completionHandler(.success(response))
		}.catch { error in
			completionHandler(.failure(error))
		}
	}
```

NetworkClient.requestObject function is this
```swift
	static func requestObject<T: Object>(_ convertible: Alamofire.URLRequestConvertible, type: T.Type) -> Promise<T>  where T: Mappable {
		return Promise<T> { seal in
			APIManager
				.request(convertible)
				.validate()
				.responseObject { (response: DataResponse<T, AFError>) in
					switch response.result {
						case .success(let value):
							seal.fulfill(value)
						case .failure(_):
							print(response)
							seal.reject(ServerError.unknown)
							
					}
				}
		}
	}
```

**Comment by jshier** [comment]
Right, network requests in the background will be automatically cancelled by `URLSession` / the system unless a) they're made from a background `URLSession` (which Alamofire doesn't yet support) or b) you've taken a background assertion as outlined in the dev forums thread I linked above. I suggest you implement the simpler `beginBackgroundTask` APIs and go from there.

**Comment by Hamna21** [comment]
@jshier I think there has been some miscommunication.

Our app is not in the background when we make a network request.
Clicking on the notification 
1. brings the app to the foreground. 
2. It then presents a view controller which shows a loader 
3. Then the network request is made which is then canceled.

Or is there something that I am missing?

**Comment by jshier** [comment]
I suggest you debug further to figure out whether the cancellation is coming from the underlying `URLSession` or Alamofire itself. Most likely it's the `URLSession`, as Alamofire has a special cancellation error it produces. In that case there's not much Alamofire can do, you'll have to figure out the cause of the cancellation. Usually this is due to the `Session` going out of scope, the app being backgrounded, or internal `URLSession` logic. If the view controller is directly triggering the network request, it shouldn't be in the background, but you should check just to be sure. 

As an aside, I suggest you rewrite your `Session` initialization, as it's rather strange.

```swift
static let APIManager: Session = {
  let configuration = URLSessionConfiguration.af.default
  let manager = Session(configuration: configuration, interceptor: NetworkClientRetrier())
  return manager
}()
```

First off, `timeoutIntervalForRequest` probably doesn't do what you expect. It is not the timeout of the initial connection, which is normally what that term means, but instead the timeout between additional data segments. Once a connection is established you shouldn't need to wait two minutes between bits of data.

Second, `httpMaximumConnectionsPerHost` is really rather pointless unless you've tested and see the defaults don't give you the throughput you need. Unless `URLSession`'s automatic behavior is too slow you really shouldn't touch this.

Third, `let delegate = Session.default.delegate` is grabbing a reference to global `Session.default`'s `delegate` instance. You really, really don't want to do this. In fact, a reference to the delegate is completely unnecessary unless you're providing your own customized instance.

In the end even the `URLSessionConfiguration` is unnecessary and you can just go with `Session(interceptor: NetworkClientRetrier())`.

**Comment by Hamna21** [comment]
@jshier 
1. How do you suggest we debug further as we are not engaging with `URLSession` directly.
2. We have a `Home controller` and then we present another controller(`View Details Controller`) which then makes an API call. This means the app is not in the background. Surprisingly the API calls made on the `Home Controller`(which is the parent) all succeed but the call made via `View Details Controller` sometimes fails.
3. Our initial code was the same as you just suggested but we added additional logic based on suggestions via StackOverflow. Reverting the code to your suggestions still leads to the issue at hand.

**Comment by jshier** [comment]
As Alamofire is just wrapping `URLSession`, so you are engaging `URLSession`. So you can set breakpoints in the relevant `SessionDelegate` methods. You could also set a breakpoint in `Request.cancel()` to see if anything is calling cancel directly, but that's unlikely. I would also check to see if you have any code that tries to interact directly with the `URLSession` underlying the `Session` instance you're using. It's possible directly cancel `URLSessionTask`s in a way that Alamofire can't see to issue its custom cancellation error and would look like this.

**Comment by Hamna21** [comment]
Also the state of the manager at the time of the failure:

<img width="635" alt="image" src="https://user-images.githubusercontent.com/20511422/176302048-26e94048-61a4-4800-81ee-e316c57240f3.png">
